cordova-serve/src/server.js

```diff
@@ -0,0 +1,137 @@
+/**
+ Licensed to the Apache Software Foundation (ASF) under one
+ or more contributor license agreements.  See the NOTICE file
+ distributed with this work for additional information
+ regarding copyright ownership.  The ASF licenses this file
+ to you under the Apache License, Version 2.0 (the
+ "License"); you may not use this file except in compliance
+ with the License.  You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing,
+ software distributed under the License is distributed on an
+ "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+ */
+
+var fs     = require('fs'),
+    http   = require('http'),
+    url    = require('url'),
+    path   = require('path'),
+    Q      = require('q'),
+    stream = require('./stream');
+
+/**
+ * @desc Launches a server with the specified options and optional custom handlers.
+ * @param {{root: ?string, port: ?number, urlPathProcessor: ?function, streamHandler: ?function, serverExtender: ?function}} opts
+ * @returns {*|promise}
+ */
+module.exports = function (opts) {
+    var deferred = Q.defer();
+
+    var root = opts.root;
+    var port = opts.port || 8000;
+
+    var server = http.createServer(function (request, response) {
+        function do404() {
+            console.log('404 ' + request.url);
+            response.writeHead(404, {'Content-Type': 'text/plain'});
+            response.write('404 Not Found\n');
+            response.end();
+        }
+
+        function do302(where) {
+            console.log('302 ' + request.url);
+            response.setHeader('Location', where);
+            response.writeHead(302, {'Content-Type': 'text/plain'});
+            response.end();
+        }
+
+        function do304() {
+            console.log('304 ' + request.url);
+            response.writeHead(304, {'Content-Type': 'text/plain'});
+            response.end();
+        }
+
+        function isFileChanged(path) {
+            var mtime = fs.statSync(path).mtime,
+                itime = request.headers['if-modified-since'];
+            return !itime || new Date(mtime) > new Date(itime);
+        }
+
+        var urlPath = url.parse(request.url).pathname;
+
+        var filePath;
+        if (opts.urlPathProcessor) {
+            var result = opts.urlPathProcessor(urlPath, request, response, do302, do404);
+            if (!result) {
+                return;
+            }
+            filePath = result.filePath;
+        }
+        if (!filePath) {
+            if (!root) {
+                throw new Error('No server root directory HAS BEEN specified!');
+            }
+            filePath = path.join(root, urlPath);
+        }
+
+        fs.exists(filePath, function (exists) {
```

Comment:
Yeah, good suggestion. It's not quite as simple as that, since a custom `urlPathProcessor()` might be provided but not want to process a particular path and fall back on default logic. I'll take a look, though, because I'm not happy with how this works currently (the logic of what `urlPathProcessor()` returns is too weird).